### ISSUE TITLE
E_ERROR: Invalid serialization data for DateTime object

### ISSUE THREAD

**Issue by mente** [igbinary/igbinary#59]
Hello!

Happens only for DateTime object. Error is not catchable and hardly reproducible (~200 errors out of 100 thousands attempts per day). Serialized data comes from memcached (that uses default serializer) and then deserialized. I will try to find proper reproduce case but until then maybe you have some ideas?

Thanks,
Alex

**Comment by TysonAndre** [comment]
@mente - Is this php 5 or php 7? (Assuming 7)


- What php version are you using? Was php upgraded around february?
- What igbinary version? (Probably irrelevant)
- What time zone/locale-related settings are used?

You can catch errors in php 7 - See https://secure.php.net/manual/en/error.getmessage.php#refsect1-error.getmessage-examples

Pseudocode for how to catch this is:

```
try {
   $unserialized = $memcached->get($key);
} catch (Error $e) {
   my_log_error("Invalid key #1 : error=#2", $key, $e->getMessage());
   if (!preg_match('/Invalid serialization data/, $e->getMessage())) {
     throw $e;  // Rethrow errors unrelated to old serialize data
   }
   // If you are done investigating, delete the key
   // $memcached->delete($key)
}
```

The hash probably contains values of "date", "timezone_type", or "timezone" (Perhaps different php versions were used to serialize, or different time zones are available on different servers).

- See https://github.com/php/php-src/blob/PHP-7.0.12/ext/date/php_date.c#L2730 and https://github.com/php/php-src/blob/PHP-7.0.12/ext/date/php_date.c#L2829
- This is not fatal: https://3v4l.org/T0q8X

Then use a memcached command line client (or client in another language) to get the raw binary data for that key (should begin with "\x00\x00\x00\x02")

**Comment by TysonAndre** [comment]
Also, I don't think there's anything special about the igbinary unserializer vs the standard unserializer for memcached for the most recent igbinary commits (May have been different in February)

**Comment by kalifg** [comment]
I am having this same issue, although we are using Redis as well as memcached.  I have captured one of the cache entries [here](https://gist.github.com/kalifg/23b1a0b995874388994ca0ce67445ef4).  This was a key on Redis

Early on the second line you can see `DateTime` and then it's followed by these 4 bytes: `14 00 11 03`

The server is running PHP 5.6 (PHP 5.6.17 (cli) (built: Jan 27 2016 15:32:44)).

The occurrence rate is similar to what the poster described, except that it clusters...so far about once a week.  The date in source data that this cache is drawn from looks fine.

**Comment by TysonAndre** [comment]
@kalifg has a strange example - this is a DateTime serialized with 0 properties (`\x14\x00`)
In php7, it's catchable, but in php 5 it isn't. 

```php
var_export(igbinary_unserialize("\x00\x00\x00\x02" .
"\x17\x08\x44\x61\x74\x65\x54\x69\x6d\x65\x14\x00"));
Fatal error: Invalid serialization data for DateTime object in php shell code on line 1
```

There's 2 possibilities

1. Igbinary couldn't serialize date, timezone, timezone_type, and silently skipped over them. That seems impossible.
2. Somehow, a DateTime was created with no properties. That also seems impossible (Maybe if time zone settings differ?)

Possible workarounds:

1. Manually calling igbinary_serialize, igbinary_unserializing, and using $redis->set(string $key, string $serialized) with no serializer.
   Failing if the string `"\x17\x08\x44\x61\x74\x65\x54\x69\x6d\x65\x14\x00"` is seen
2. In `igbinary_unserialize_object` of src/php5/igbinary.c, add something similar to this patch (not tested/compiled, not common enough of an issue to be on master:

```php
2250     if (r) {
2251         return r;
2252     }
2253 
+ if (name_len == 8 && strcasecmp(name, "DateTime") == 0 && zend_hash_num_elements(Z_OBJPROP_P(z)) == 0) {
+    zend_error(E_WARNING, "igbinary_unserialize_object: Empty DateTime, position %zu", igsd->buffer_offset));
+    return 1;
+ }
2254     if (Z_OBJCE_PP(z) != PHP_IC_ENTRY && zend_hash_exists(&Z_OBJCE_PP(z)->function_table, "__wakeup", sizeof("__wakeup"))) {
2255         INIT_PZVAL(&f);
```

**Comment by kalifg** [comment]
Thank you for replying so quickly!

After doing a little more research on my own, I think that we are both suffering from this [bug](https://bugs.php.net/bug.php?id=72011) in PHP, which only shows itself under high load.  It is present in both 7.x and 5.x; here's an summary from the ticket:

> The bug occurs when the max_execution_time is reached *while* GC is running. Because of the way the timeout is handled, GC is never given an opportunity to clean up and reset the gc_active flag or finish collecting garbage. 
> 
> In PHP-FPM, the worker does not exit when the timeout error occurs, but it doesn't clean up the invalid GC state either. When the bug has occurred in an FPM worker process, the worker continues serving requests, but any DateTime objects created by the worker appear to have no properties when accessed via the get_properties handler. Additionally, a bugged FPM worker will segfault if the bug test code is run twice. I'm not sure why that is happening, but it's not surprising, as once the bug occurs, the worker is in an invalid and indeterminate state anyway.

It seems that it is indeed not an igbinary-specific issue.

There is a [fix](https://github.com/php/php-src/commit/248fdfcf7356f2c20ab1e6afd1e9f295d08331c7), tagged for 7.1.0beta3 and 5.6.24RC1.

**Comment by TysonAndre** [comment]
So the fix is already released - the most recent releases of php are 7.1.0RC5 and 5.6.27. http://php.net/downloads.php#v5.6.27
